aten/src/ATen/FunctionalTensorWrapper.cpp

```diff
@@ -0,0 +1,366 @@
+
+#include <ATen/FunctionalTensorWrapper.h>
+
+#include <ATen/FunctionalInverses.h>
+#include <ATen/WrapDimUtils.h>
+#include <ATen/core/LegacyTypeDispatch.h>
+#include <c10/util/Exception.h>
+
+#include <c10/util/irange.h>
+
+namespace at {
+
+FunctionalTensorWrapper::FunctionalTensorWrapper(Tensor value)
+  : c10::TensorImpl(
+      c10::Storage(c10::make_intrusive<functionalization::FunctionalStorageImpl>(value.device(), value.numel(), value.dtype())),
+      c10::DispatchKeySet(DispatchKey::Functionalize),
+      value.dtype()
+    ),
+    value_(value),
+    // Note: "level" is a concept that we don't know how to compute in core.
+    // For now I'm retroactively setting this in functorch,
+    // but once Open Multiple Dispatch lands we should be able to calculate this in core.
+    level_(-1)
+{
+  // shallow_copy_from overwrites the storage and dispatch keyset...
+  auto functional_storage = storage_;
+  shallow_copy_from(value.getIntrusivePtr());
+  storage_ = functional_storage;
```

Comment:
Hmm, you sure you want to use `shallow_copy_from` to implement this then?

Reply:
i figured manually fixing up these 3 fields was better than either manually writing out the other 15 that I need, or adding a new TensorImpl method specifically for this use case :(

If you have a better suggestion, happy to follow it